job-manager: support events journal

chu11 opened the pull request:
Built on top of PR #3254.  It's the final in my series of job-manager / job-info refactoring / changes.

A few notes (mostly same notes as my prior attempted PR on this, #3238)

- introduces a `seq` into the wrapped entries, so listeners can determine if they have seen the event for a job before (via the KVS).

- b/c the `annotations` object of a job can change, we have to  `json_deep_copy()` job annotations that go into the journal so they can be preserved.

- i made the maximum journal length configurable.  The default is 1000 entries, which translates to 100-125 jobs on average (8 entries for a typical successful job, not counting any extra annotations that may come from the scheduler).  Dunno if this is too small?  Would we want longer?

- b/c of the journaling, I don't really need to synchronize on the `job-manager.events.listeners` stat anymore.  I could remove the stat, but left it in there for the time being since it doesn't hurt to have it there.

garlick commented:
In this series, the job manager journal feature is kinda spread out across the PRs.  I might have not been clear about what I was hoping for when we discussed it earlier, but I had thought it could be implemented as a standalone job manager feature with tests that don't involve job-info, in one PR.  I was also hoping it could be isolated from event.c maybe in journal.c or something, following the pattern of other "services" within the job manager.

I though the series would consist of something like
* job-manager: implement journal of job events
* job-info: use job manager journal for job annotations
* job-info: use job manager journal instead of published state transitions

etc

I haven't had a chance to look very closely so maybe I'm just getting confused - apologies if so!

chu11 commented:
> In this series, the job manager journal feature is kinda spread out across the PRs. I might have not been clear about what I was hoping for when we discussed it earlier, but I had thought it could be implemented as a standalone job manager feature with tests that don't involve job-info, in one PR. I was also hoping it could be isolated from event.c maybe in journal.c or something, following the pattern of other "services" within the job manager.

It can definitely be split out with some refactoring.  But perhaps we should talk on Monday before I go down that route to make sure we're on the same page.

> I though the series would consist of something like
>    job-manager: implement journal of job events
>    job-info: use job manager journal for job annotations
>    job-info: use job manager journal instead of published state transitions

Hmmm.  I can see your train of thought.  I think I thought of it more as "limiting how painful will it be for me to convert job-info to use the events stream", so I took a series of steps to make that path easier.  There were a lot of hidden steps along the way that i eventually merged into fewer commits :P

Hopefully the end is close to what you imagined (minus a refactoring perhaps).

chu11 commented:
@garlick now that i think about it, i suppose with the series of PRs now complete (perhaps minus a refactoring), it would perhaps be possible to re-order to the series you imagine.  or even into one giant PR.  although i think it would be a lot to review.

garlick commented:
I'll go ahead and check out the end product and try to give some feedback on the job manager end while we discuss.

It would be nice to have the job manager journal split out in its own PR rather than spread across multiple PRs, but I don't want to make a bunch of extra work for you.  Let's see, right now we have:

#3236 job-manager: Move annotations updates to `job-manager.events`, remove `job-annotations` event
#3253 job-manager: send events stream immediately 
#3254 job-info: refactor to use only event stream, not job-state transitions
#3261 job-manager: support events journal

Maybe the three job manager PRs could be simply be combined into one?

garlick commented:
I'd really like to see the job manager journal broken out to its own journal.c source file.  I think the main thing getting in the way of hat is that the journal is still synchronized with the KVS eventlog batching so that events emitted to the latter land first.  Is that still necessary?  If a user of the journal has received an event from the journal, does it need to also fetch it from the KVS?  I think we established that synchronization on R and jobspec are already a step removed from the job manager so submit and alloc events still land in the journal _after_ those objects are commited.

chu11 commented:
> #3236 job-manager: Move annotations updates to job-manager.events, remove job-annotations event
> #3253 job-manager: send events stream immediately
> #3254 job-info: refactor to use only event stream, not job-state transitions
> #3261 job-manager: support events journal

> Maybe the three job manager PRs could be simply be combined into one?

Well #3236 is really `job-manager` + `job-info`, since its the removal of the old `job-annotations` event and its removal will break `job-info` and tests.  I viewed that one as independent of the others since its less about the events journal and more about the removal the `job-annotations` event.

I guess I could leave removal of the `job-annotations` event after this in another PR?  But it'll be a nice chunk of work.

Thinking about it, I think #3253 & #3261 could probably be combined into one PR w/o breaking `job-info` (assuming #3236 goes in first), leaving #3254 in a different PR, which could be merged afterwards.

chu11 commented:
> Edit: I think I misread the code and the journal does get the updates right away, so there should be no impediment to decoupling from the KVS batching logic and state transition pub?

Yes, the journal gets the updates right away.  the journal events are sent out immediately on each batch cycle and not dependent on the KVS.

The main reason that the journal code is in the events code is b/c A) journal is updated with `event_job_post_pack()` (and side functions) and B) the listener responses are sent the responses in batches, so I'm piggybacking the batching code.

Are you imagining the journal events are sent immediately and not in batch form?  Just trying to think of how to split this out into its own `journal.c` file right now.

garlick commented:
OK, why don't we try to get #3236 merged first and we can go from there.  That looks reasonable.

garlick commented:
> Are you imagining the journal events are sent immediately and not in batch form? Just trying to think of how to split this out into its own journal.c file right now.

What's the reason for batching the responses?  I assumed it was just leftover from when the KVS and journal were synchronized.

chu11 commented:
> What's the reason for batching the responses? I assumed it was just leftover from when the KVS and journal were synchronized.

Just to limit the number of RPCs in flight.  i.e. do we send one RPC for each event that occurs, or shall we batch up a few into one RPC.

garlick commented:
That might save the job manager and job info some envelope overhead in the high throughput case, but honestly I'd prioritize having a well factored journal feature over that at least until it becomes a problem.

chu11 commented:
> That might save the job manager and job info some envelope overhead in the high throughput case, but honestly I'd prioritize having a well factored journal feature over that at least until it becomes a problem.

Hmmm.  Due to the fact that the events stream began in the batching code all the way back in #3208, we send out multiple events wrapped inside an array.  If we go with a event per RPC model, the `job-manager` change would break `job-info`.  Thus:

```
#3253 job-manager: send events stream immediately
#3254 job-info: refactor to use only event stream, not job-state transitions
#3261 job-manager: support events journal
```

would all have to go into one PR.  Do we want to go down that path?

garlick commented:
I was assuming the protocol wouldn't change.  E.g. if we're wrapping events in an array per response, keep doing that.  Then we won't have to change the client later if we re-add a batching mechanism.  I just don't think it's necessarily a great architecture to have it tacked on the KVS commit batching since it doesn't need to be synchronized to it (unlike the published state transitions).

chu11 commented:
> I was assuming the protocol wouldn't change. E.g. if we're wrapping events in an array per response, keep doing that. Then we won't have to change the client later if we re-add a batching mechanism. I just don't think it's necessarily a great architecture to have it tacked on the KVS commit batching since it doesn't need to be synchronized to it (unlike the published state transitions).

Without batching and sending only 1 event per RPC, the array is sort of pointless.  But I guess we can just refactor that out at the end in a separate commit.

garlick commented:
Either way.  It wouldn't hurt to leave it in IMHO if we think the natural queuing provided by the outbound socket will be much higher overhead than a separate queue with batching.

chu11 commented:
re-pushed, per discussion above still built on top of #3236.  This PR solves both #3245 and #3172.  Quick summary of what is done in this PR (including some re-hash from above, just as a reminder).

- First solved #3245 by simply having events sent to listeners immediately.  The sending of events is completely divorced from the event batching that was in place before.

- introduces a `seq` into the wrapped entries, so listeners can determine if they have seen the event for a job before (via the KVS).

- b/c the `annotations` object of a job can change, we have to  `json_deep_copy()` job annotations that go into the journal so they can be preserved.

- i made the maximum journal length configurable.  The default is 1000 entries, which translates to 100-125 jobs on average (8 entries for a typical successful job, not counting any extra annotations that may come from the scheduler).  Dunno if this is too small?  Would we want longer?

- the last few commits are all renaming "cleanup".  I changed the `job-manager.events` topic to `job-manager.events-journal`, the `events.listener` stat to `journal.listener`, and renamed files / tests names / etc.  Typical cleanup that had to "cross" modules / tests / etc.

- b/c of the journaling, I don't really need to synchronize on the `job-manager.journal.listeners` stat anymore.  I could remove the stat, but left it in there for the time being since it doesn't hurt to have it there.

chu11 commented:
re-pushed, think there was a racy test, rebased while i was at it

chu11 commented:
re-pushed, fixing up given changes in #3236

garlick commented on src/modules/job-manager/annotate.c:
```diff
@@ -145,19 +146,27 @@ void annotate_handle_request (flux_t *h,
     }
     if (annotations_update (ctx->h, job, annotations) < 0)
         goto error;
+    if (job->annotations) {
+        if (!(tmp = json_deep_copy (job->annotations)))
+            goto nomem;
+    }
     if (event_job_post_pack (ctx->event,
                              job,
                              "annotations",
                              EVENT_JOURNAL_ONLY,
                              "{s:O?}",
-                             "annotations", job->annotations) < 0)
+                             "annotations", tmp) < 0)
         goto error;
```
This might not be a comment that should be addressed in this PR, but should the annotation event be a snapshot of all annotations?  It seems like it should be for a specific annotation.  The journal consumer would then apply each event and build the same snapshot.  That would seem more consistent with the information that other events carry.  For example, we don[t tell you the current job state in a state transition.  You have to apply each event and implement the state machine to recreate the state from the events.

chu11 replied:
I assume you mean a snapshot of all annotations for the job?  Seems like a good idea. Perhaps should be rolled in for the discussion in #3081, as part of the general "persistence" of annotations.

garlick replied:
To be clear, my idea was that an annotation event would _not_ be a snapshot.  Instead it would be something along the lines of a single key-value pair.

chu11 replied:
oh i misread your comment.  I see your point now, effectively each event in the journal should be the "delta" from the prior annotations.  That does make sense.  Perhaps a cleanup for a follow up PR.

garlick commented on src/modules/job-manager/journal.h:
```diff
@@ -0,0 +1,46 @@
+/* Process the event as needed, such as incrementing job sequence
+ * number, and send to any listeners that request the event.
+ */
+int journal_process_event (struct journal *journal,
+                           struct job *job,
+                           int *seqp,
+                           const char *name,
+                           json_t *entry);
```
I'd probably call this something like `journal_append_event()`.

Would it be OK to just pass in the job ID rather than the whole job struct?

I asked about the sequence number in an earlier comment.  If we could nix that it would be nice.

chu11 replied:
> Would it be OK to just pass in the job ID rather than the whole job struct?

Looks that'll work now given all the changes.  Good catch.

chu11 replied:
hmmm, to me `journal_append_event()` suggests we're only appending to the journal list, but the function also sends to listeners.  a better name isn't coming to mind.  I considered `journal_send_event()` but it similarly suggests just sending.  We could of course be dumb and call it `journal_send_and_append_event()`.

garlick replied:
Well I was thinking that the abstract model is that the journal is a queue that is added to by the job manager and  consumed by users, so from the job manager perspective, "append" would be OK.  For example, in the future we might want to use a prep/check/idle pattern inside journal.c to wake up consumers after new events have been added to the journal.

No big deal though - it's an internal function and I'm fine if we leave it.

garlick reviewed (commented):
OK, I think this looks pretty good.  I did have some questions inline about the job sequence number.  If that's mitigation for job-info getting events from multiple sources, could we possibly nix that here?  Feels like it should be unnecessary.

I think moving the journal code to journal.c really helps with the casual grokking.

I would approve but wanted to hear back on the id first.

garlick commented on src/modules/job-manager/alloc.c:
```diff
@@ -227,12 +228,16 @@ static void alloc_response_cb (flux_t *h, flux_msg_handler_t *mh,
         if (annotations_update (h, job, annotations) < 0)
             flux_log_error (h, "annotations_update: id=%ju", (uintmax_t)id);
         if (annotations) {
+            if (job->annotations) {
+                if (!(tmp = json_deep_copy (job->annotations)))
+                    goto nomem;
+            }
```
Er never mind, I get it.  See question about content of annotation event.

chu11 replied:
I should perhaps add a comment

chu11 commented on src/modules/job-manager/event.c:
```diff
@@ -521,14 +521,15 @@ static int get_timestamp_now (double *timestamp)
- * the eventlog entry in another object with the job id
+/* we need to send the job id & seq along with each eventlog entry, so
+ * wrap the eventlog entry in another object with the job id & seq
  */
```
Sounds good.  FYI per discussion in #3209 

```
Better would be to include a per-job event sequence number in the journal event envelope (where we are adding the job id). During KVS replay, both producer (job manager) and consumer would count how many events are applied to each job. Then, for each job, events in the journal could be discarded if they contain an earlier sequence number. This would also provide a way to detect if events were missed, e.g. if the journal buffer was overrun at the source.
```

garlick replied:
If that's the case don't we need to increment job->seq in restart.c?  (Or rather `job_create_from_eventlog()` in job.c?)  Your comment reminded me that the main use case for this is when both job-info and job-manager are restarting and are creating initial state from the KVS.  There could be some synchronization issues there at start up.

Also would be good in that comment to say something about the fact that the sequence does not count JOURNAL_ONLY events.

chu11 replied:
> If that's the case don't we need to increment job->seq in restart.c?

Oh crud!  You're right, I had totally forgotten that.  I did it in the `job-info` side (follow up PR), but not on the `job-manager` side.

chu11 commented:
rebased & re-pushed per comments above, except for the  renaming of `journal_process_event()`.

chu11 commented:
re-pushed correcting the fact that `job->seq` was not incremented during the initialization of jobs via the KVS.   There were several ways to do this, some of which were uglier than I would have liked.

I elected to refactor things and increment `job->seq` in `event_job_update()`.  This was a pretty clean change, but does require the caller to call `journal_process_event()` before `event_job_update()`, which may not be the most obvious thing to do.  I added comments as such.

garlick commented:
That seems OK, except if `event_job_update()` fails, no event is posted but the sequence number is still incremented.  The increment should probably be moved all the way to the end of that function, right before it returns success.

It might actually be a little bit clearer, since that function advertises its purpose as advancing the state machine, to just increment `job->seq` directly in the three places where `event_job_update()` is called.

Would it be slgihtly clearer if `job->seq` was renamed to `job->eventlog_count`?  "seq" is a bit generic here.

I noticed there's a little style inconsistency that's crept in in a few places
```
   if (EVENT_JOURNAL_ONLY & flags)
        goto out;
```
for consistency with the rest of the code base, we should probably flop that to `(flags & EVENT_JOURNAL_ONLY)`.

There used to be a version of gcc that complained about using a bitwise operator in a logical expression without parens.  If you're visiting those places anyway, you could add a second set, e.g. `((flags & EVENT_JOURNAL_ONLY))`, or if that's off base and maybe I dreamed about that version of gcc, just ignore me :-)

chu11 commented:
rebased & re-pushed fixing up the comments from @garlick 

- moved the `job->seq++` out of `event_job_update()`, and just did it in within the 3 locations `event_job_update()` is called.

- I renamed `seq` to `eventlog_seq`, both variable names and the field in the journal event.  Adjusted some comments / tests accordingly for name change.

- added a cleanup commit for the `EVENT_JOURNAL_FLAG & flags`

garlick commented on src/modules/job-manager/event.c:
```diff
@@ -601,18 +614,25 @@ int event_job_post_pack (struct event *event,
     int saved_errno;
     double timestamp;
     flux_job_state_t old_state = job->state;
+    int eventlog_seq = EVENT_JOURNAL_ONLY & flags ? -1 : job->eventlog_seq;
```
I think you missed flipping flags and constant, and probably should also put parens around the logical expressin.

garlick reviewed (approved):
I this is good to go after the one minor tweak mentioned mentioned just now.  Thanks!

chu11 commented:
rebased & re-pushed fixing the nit @garlick found.  LMK if its ok to set MWP

garlick commented:
Yes go ahead!
